### DISCUSSION TITLE
CefSharp browser css rendering not working

### DISCUSSION THREAD

**Discussion by rock117** [cefsharp/CefSharp#3536] [Q&A]
Hello CefSharp# team,

My Company's project is a WPF base APP, recently we plan to integrate CefSharp into our APP.

We have UI setting such as color, font-size in WPF dialog, once user change those settings, it should apply the settings to cefsharp browser's  component(by calling ExecuteScriptAsync which define js function to change DOM's style in JS side).

Our APP is like PPT, left side is the navigation, and the right side is the details page, when switch the page  in left side, the right show the corresponding page(rendered in cefsharp browser).

If user change the UI setting in WPF dialog, all pages' browser component need to apply the style.

Everything goes well if there are one or two pages in the left, the browser rendered correctly, see below, the text color is ok
![image](https://user-images.githubusercontent.com/2640825/117009363-3d0fc580-ad1e-11eb-9b11-0f38b3d21711.png)

but if there are more pages, the browser not render css correctly, see below
![image](https://user-images.githubusercontent.com/2640825/117009591-819b6100-ad1e-11eb-9de3-359a7ff02beb.png)
Even if I change the color manually in the Devtools, the browser still not apply

**Comment by amaitland** [comment]
What version are you using? How does Chrome behave if you change the css using DevTools?

**Reply by rock117** [reply]
hi @amaitland , My Version is v88.2.90, and I had tried v89.0.170, it also not worked.  My PC OS is Windows 10.
If I changed the css in DevToos, The browser does not render, I had break point in the OnPaint method(a CustomBrowser overide method, see below), but the code does not execute.

![image](https://user-images.githubusercontent.com/2640825/117090543-310e1d00-ad8b-11eb-8d0d-0660bf21ec7c.png)

But if I export the browser as PDF using browser.PrintToPdfAsync, the pdf file apply the css correctly.

**Reply by rock117** [reply]
After added the following code, the bug fixed
` if (!Cef.IsInitialized)
            {
                var setting = new CefSettings();
                setting.SetOffScreenRenderingBestPerformanceArgs();
                Cef.Initialize(setting);
                Cef.EnableHighDPISupport();
            }`

**Reply by amaitland** [reply]
What graphics card do you have? Have you tested on different hardware? 

If you have an ancient GPU then disabling your GPU is fairly common. If you have an Intel Iris Xe Integrated GPU(11th Gen) there are known issues with that (See release notes for additional details if you are using that GPU)

**Reply by rock117** [reply]
@amaitland ,  the actual code which fixed the bug is setting.SetOffScreenRenderingBestPerformanceArgs(). the DPI has no affect on it.
Currently, I only test in my PC,  the graphics card is

![image](https://user-images.githubusercontent.com/2640825/117105332-5f522380-adb0-11eb-99cc-5ec941b41547.png)

**Reply by amaitland** [reply]
See https://github.com/cefsharp/CefSharp/blob/09652614bd65408ee255efcbf9db7048df775182/CefSharp.Core/CefSettingsBase.cs#L427

Calling SetOffScreenRenderingBestPerformanceArgs disables GPU acceleration and GPU compositing. 

Try just disabling GPU compositing to start with. 

Loading chrome://gpu in Chrome might give you some hints, if GPU acceleration or compositing are disabled.
